pkg/controllers/sidecar/sync.go

```diff
@@ -2,41 +2,67 @@ package sidecar
 
 import (
 	"context"
-
-	"github.com/scylladb/scylla-operator/pkg/util/network"
+	"github.com/scylladb/scylla-operator/pkg/scyllaclient"
 
 	"github.com/pkg/errors"
 	"github.com/scylladb/scylla-operator/pkg/controllers/cluster/util"
 	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/util/network"
 	corev1 "k8s.io/api/core/v1"
 )
 
-func (mc *MemberReconciler) sync(ctx context.Context, memberService *corev1.Service) error {
+func (mc *MemberReconciler) sync(ctx context.Context, memberService *corev1.Service) (bool, error) {
 	// Check if member must decommission
 	if decommission, ok := memberService.Labels[naming.DecommissionLabel]; ok {
 		host, err := network.FindFirstNonLocalIP()
 		if err != nil {
-			return errors.Wrapf(err, "error during decommission")
+			return true, errors.Wrapf(err, "Can't get node ip")
 		}
 		// Check if member has already decommissioned
 		if decommission == naming.LabelValueTrue {
-			return nil
+			mc.logger.Debug(ctx, "Node is already decommissioned")
+			return false, nil
 		}
-		// Else, decommission member
-		if err := mc.scyllaClient.Decommission(ctx, host.String()); err != nil {
-			mc.logger.Error(ctx, "Error during decommission", "error", errors.WithStack(err))
+		opMode, err := mc.scyllaClient.OperationMode(ctx, host.String())
+		if err != nil {
+			return true, errors.Wrapf(err, "error on getting node operation mode: %s", errors.WithStack(err))
 		}
-		// Confirm memberService has been decommissioned
-		if opMode, err := mc.scyllaClient.OperationMode(ctx, host.String()); err != nil || !opMode.IsDecommisioned() {
-			return errors.Wrapf(err, "error during decommission, operation mode: %s", opMode)
+		switch opMode {
+			case scyllaclient.OperationalModeLeaving:
+				// If node is leaving, keep retrying while it left the cluster
+				return true, nil
+			case scyllaclient.OperationalModeDraining:
+				// If node is draining, keep retrying while it fully drained the cluster
+				return true, nil
+			case scyllaclient.OperationalModeDrained:
+				return true, errors.Errorf("Can't decommission node which is in DRAINED state, node needs to be restarted")
+			case scyllaclient.OperationalModeNormal:
+				// Decommission node only if it is in normal mode
+				if decommissionErr := mc.scyllaClient.Decommission(ctx, host.String()); decommissionErr != nil {
+					// Decommission is long running task, so request fails due to the timeout in most cases
```

Comment:
Shouldn't Scyla enter Leaving state after first request? Condition below suggest it does so this branch shouldn't be picked up on retry.
If next reconcilation happens too fast, we have control over it. 
https://github.com/scylladb/scylla-operator/blob/bc214d9b9dbb7462bbd436ae393e288d50fc386a/vendor/sigs.k8s.io/controller-runtime/pkg/reconcile/reconcile.go#L32

Reply:
We can also check if timeout happened or it was some other failure and react based on error value

Reply:
I did not find that it does so.

Problem here that retrying is done on scyllaclient level, so sync is not getting 'control' back from mc.scyllaClient.Decommission(ctx, host.String()) up until scyllaclient exhausted all retrys.

Even if we get scyllaclient fixed, decommission can take unpredictable amount of time, so best thing to do here would be relaying on node status, it if is in decommissioned or decommissining state, then we got what we wanted.
Additionaly we cover all possible cases that could happen due to the race condition with nodetool commands.